core/isolate.rs

```diff
@@ -182,17 +191,34 @@ impl Isolate {
     let isolate = unsafe { Isolate::from_raw_ptr(user_data) };
     let control_shared = isolate.shared.shift();
 
-    let op = if control_argv0.len() > 0 {
+    let (op, promise_id, is_sync) = if control_argv0.len() > 0 {
       // The user called Deno.core.send(control)
       if let Some(ref f) = isolate.config.dispatch {
-        f(control_argv0.as_ref(), PinnedBuf::new(zero_copy_buf))
+        let promise_id =
+          i32::from_ne_bytes(control_argv0[..4].try_into().unwrap());
```

Comment:
I don't like that the control buffer is no longer opaque to core. This constraints how op implementors do serialization.

There are a couple of distinct interface problems that are trying to be achieved here:
1. Allow `Op` to be generic on the Error.
1. Pass `is_sync` into the dispatch function.

I think the first item is achieved. Consider submitting it as a separate PR, I think we can land that. (Although I have a question about the performance of converting `Op::Async<E>` to `Op::Async<()>`, it seems the `Box<Future>` must be unwrapped to handle the error and serialize further.)

Regarding `is_sync`: 

I much prefer how you had it before, where `is_sync` (or `op_id`) actually becomes a parameter of `Deno.core.dispatch`. Having core inspect the payload of buffers is breaking the abstraction layer.

There is an incongruence between the op request and response: the JS side could as for a sync op and receive an async op. Or vise-versa. You've solved this by adding asserts, but if we more deeply refactor this interface we maybe could remove the asserts and solve the problem of having to add a parameter to `Deno.core.dispatch()`.

What if the requester does not specify if it wants sync or async? The requester somehow must encode what they desire into the control buffer - and the op creator returns whatever. When the JS side reads the response, it determines how to deal with it. This could be achieved with the simple convention: if `Deno.core.dispatch()` returns `null`, then the op is async. If not, the JS side can assume it is sync. The JS side can then do the asserts to check that the response matches what they requested, and core is not effected.

Reply:
> There is an incongruence between the op request and response: the JS side could request a sync op and receive an async op. Or vise-versa. You've solved this by adding asserts on the Rust side, but if we more deeply refactor this interface we maybe could remove the asserts and solve the problem of having to add a parameter to Deno.core.dispatch().

Maybe if the rust side responds with a promise id then we know to expect a promise, and if it's not a number(a buffer) then we know the op is complete. Then it really doesn't matter how we respond unless the js side is expecting something specific, and in that case just encode it into the control.

Reply:
Is it necessary to know the promise id?